From: =?utf-8?q?Maxime_B=C3=A9lair?= <maxime.belair@canonical.com>
Date: Wed, 8 Apr 2026 17:39:08 +0200
Subject: [PATCH] snap_browser: Adding missing perms when opening from an
 hyperlink
MIME-Version: 1.0
Content-Type: text/plain; charset="utf-8"
Content-Transfer-Encoding: 8bit

On pdf viewers (evince, papers), opening the default navigator (Firefox,
Opera, ...) through clicking an hyperlink, was crashing due to missing
rules. Adds the extra rules needed to allow this behavior.

Fixes: https://bugs.launchpad.net/ubuntu/+source/papers/+bug/2127874

Signed-off-by: Maxime Bélair <maxime.belair@canonical.com>
---
 profiles/apparmor.d/abstractions/snap_browsers | 23 ++++++++++++++++++++++-
 1 file changed, 22 insertions(+), 1 deletion(-)

diff --git a/profiles/apparmor.d/abstractions/snap_browsers b/profiles/apparmor.d/abstractions/snap_browsers
index 301efe6..d5a1add 100644
--- a/profiles/apparmor.d/abstractions/snap_browsers
+++ b/profiles/apparmor.d/abstractions/snap_browsers
@@ -17,6 +17,7 @@ profile snap_browsers {
   /{,snap/core/[0-9]*/,snap/snapd/[0-9]*/}usr/lib/snapd/snapd r,
   /{,snap/core/[0-9]*/,snap/snapd/[0-9]*/}usr/lib/snapd/snap-seccomp rPix,
   /{,snap/core/[0-9]*/,snap/snapd/[0-9]*/}usr/lib/snapd/snap-confine Pix,
+  /{,snap/core/[0-9]*/,snap/snapd/[0-9]*/}usr/lib/@{multiarch}/* mr,
   /var/lib/snapd/system-key r,
   /run/snapd.socket rw,
 
@@ -32,10 +33,30 @@ profile snap_browsers {
   dbus send bus="session" path="/org/freedesktop/systemd1" interface="org.freedesktop.systemd1.Manager" member="StartTransientUnit" peer=(name="org.freedesktop.systemd1"),
   dbus receive bus="session" path="/org/freedesktop/systemd1" interface="org.freedesktop.systemd1.Manager" member="JobRemoved",
 
-  /sys/kernel/security/apparmor/features/ r,
+  /sys/kernel/security/apparmor/features/{,**} r,
 
   # allow launching official browser snaps.
   /snap/{brave,chromium,firefox,opera}/[0-9]*/meta/{snap.yaml,hooks/} r,
   /var/lib/snapd/sequence/{brave,chromium,firefox,opera}.json r,
   /var/lib/snapd/inhibit/{brave,chromium,firefox,opera}.lock rk,
+
+  owner @{run}/user/@{uid}/.mutter-Xwaylandauth.* r,
+  owner @{run}/user/@{uid}/snap.{brave,chromium,firefox,opera}/ rw,
+  /sys/fs/cgroup/cgroup.controllers r,
+
+  /usr/bin/getent rmCx,
+  profile getent /usr/bin/getent {
+    include <abstractions/base>
+    include <abstractions/nameservice-strict>
+    @{exec_path} rm,
+  }
+
+  /usr/bin/systemctl rmCx,
+  profile systemctl /usr/bin/systemctl {
+    include <abstractions/base>
+    @{exec_path} rm,
+
+    unix bind addr=@*/bus/systemctl/system,
+  }
+
 }
